ihdad/src/HDA/stream.rs

```diff
@@ -0,0 +1,308 @@
+		self.buff_length.read()
+	}
+
+	pub fn run(&mut self) {
+		let val = self.control() | (1 << 1);
+		self.set_control(val);
+	}
+
+	pub fn stop(&mut self) {
+		let val = self.control() & !(1 << 1);
+		self.set_control(val);
+	}
+	
+	pub fn stream_number(&self) -> u8 {
+		((self.control() >> 20) & 0xF) as u8
+	}
+	
+	pub fn set_stream_number(&mut self, stream_number: u8) {
+		let val = (self.control() & 0x00FFFF ) | (((stream_number & 0xF ) as u32) << 20);
+		self.set_control(val);
+	}
+
+	pub fn set_address(&mut self, addr: usize) {
+		self.buff_desc_list_lo.write( (addr & 0xFFFFFFFF) as u32);
+		self.buff_desc_list_hi.write( ( (addr >> 32) & 0xFFFFFFFF) as u32);
+	}
+
+	pub fn set_last_valid_index(&mut self, index:u16) {
+		self.last_valid_index.write(index);
+	}
+
+	pub fn link_position(&self) -> u32 {
+		self.link_pos.read()
+	}
+
+	pub fn set_interrupt_on_completion(&mut self, enable:bool) {
+		let mut ctrl = self.control();
+		if enable {
+			ctrl |= (1 << 2);
+		} else {
+			ctrl &= !(1 << 2);
+		}
+		self.set_control(ctrl);
+	}
+
+	pub fn buffer_complete(&self) -> bool {
+		self.status.readf(1 << 2)
+	}
+
+	pub fn clear_interrupts(&mut self) {
+		self.status.write(0x7 << 2);
+	}
+
+	// get sample size in bytes
+	pub fn sample_size(&self) -> usize {
+		let format = self.format.read();	
+		let chan = (format & 0xF) as usize;
+		let bits = ((format >> 4) & 0xF) as usize;
+		match bits {
+			0 => 1 * (chan + 1),
+			1 => 2 * (chan + 1),
+			_ => 4 * (chan + 1),
+		}
+	}
+}
+
+pub struct Stream {
+	buff:                   usize,
+	buff_phys:              usize,
+	buff_length:            usize,
+	output_current_block:   usize,
+	block_length:           usize,
+	block_count:            usize,
+}
+
+
+#[repr(packed)]
+pub struct BufferDescriptorListEntry {
+	addr:     Mmio<u64>,
+	len:      Mmio<u32>,
+	ioc_resv: Mmio<u32>,
+}
+
+impl BufferDescriptorListEntry {
+	pub fn address(&self) -> usize {
+		self.addr.read() as usize
+	}
+
+	pub fn set_address(&mut self, addr:usize) {
+		self.addr.write(addr as u64);
+	}
+
+	pub fn length(&self) -> u32 {
+		self.len.read()
+	}
+
+	pub fn set_length(&mut self, length: u32) {
+		self.len.write(length)
+	}
+
+	pub fn interrupt_on_completion(&self) -> bool {
+		(self.ioc_resv.read() & 0x1) == 0x1
+	}
+
+	pub fn set_interrupt_on_complete(&mut self, ioc: bool) {
+		self.ioc_resv.writef(1, ioc);
+	}
+}
+
+
+
+pub struct StreamBuffer {
+	phys:   usize,
+	addr:   usize,
+	
+	len: usize,
+}
+
+impl StreamBuffer {
+	pub fn new(length: usize) -> result::Result<StreamBuffer, &'static str> {
+		let phys = unsafe {
+			syscall::physalloc(length)
+		};
+		if !phys.is_ok() {
+			return Err("Could not allocate physical memory for buffer.");
+		}
+
+		let phys_addr = phys.unwrap();
+
+		let addr = unsafe { 
+			syscall::physmap(phys_addr, length, MAP_WRITE)
+		};
+
+		if !addr.is_ok() {
+			return Err("Could not map physical memory for buffer.");
+		} else {
+			unsafe {syscall::physfree(phys_addr, length);}
+		}
```

Comment:
You return without freeing on error, and free pages when you succeed?
I'm not sure about the semantics of `physmap` and `physfree`, but this seems like a bug.